update master school file 15-18

maggiesu0725 opened the pull request:
finally making some progress on the master school file. My next step is to geocode the schools.

maggiesu0725 committed:
update master school file 15-18

maggiesu0725 commented:
Could you take a look at D:\DCData\Libraries\Schools\Prog\Master school file\schools_15_16.sas? This is Step 3 in teh Master School File Rules and Procedures, which is to merge the new years to the master school file. It seems there are a bunch of errors but the program and errors are beyond me to debug. Could you take a look? Thanks! This seems the second to last step of updating the master school file and ideally I need it updated by 2018. @rpitingolo

maggiesu0725 commented:
@ptatian these are the programs to update master school file according to the instruction. Procedure is here https://urbanorg.box.com/s/dhw79vjc5hc7trs16og1pvepblmkfvjf

I updated and assigned new IDs for school year 2015-2018 and they are in the format for master school file updates. Saved on my D drive. I can put it on L somewhere?

ptatian commented:
@maggiesu0725 Yes, please put the ID file on L: so that I can access it. Thank you. 

Which program has the problems?

maggiesu0725 committed:
schools_15_16

maggiesu0725 commented:
saved here L:\Libraries\Schools\Data\Enrollment @ptatian  cc @rpitingolo

ptatian commented:
@maggiesu0725 I need the contents of this folder, too. `D:\DCData\Libraries\Schools\Data\Master school file\15_16_data`

rpitingolo commented:
@ptatian I think we are missing the contents of the 15_16_data folder. Step 1 says to "Download school directory data with geographic information for each new school year from the DC Public Schools and the DC Public Charter School Board web sites" but doesn't include any links. I _think_ this data has been moved over to the open data portal.

http://opendata.dc.gov/datasets/public-schools
http://opendata.dc.gov/datasets/charter-schools

These are similar to the property data in that they're not dated and the metadata says "Current for the 2018-2019 school year". Since we didn't download files over time I don't think we can specifically get past years without a special request.

maggiesu0725 commented:
@rpitingolo we got the 15-18 data from ODCA. So I'm not downloading data from anywhere. cc @ptatian

rpitingolo commented:
@maggiesu0725 can you please show us where that data is saved? I can't seem to find it anywhere on L.

maggiesu0725 commented:
L:\Libraries\Schools\Data\Enrollment\enroll_15_16 16_17 17_18 @rpitingolo

rpitingolo commented:
@maggiesu0725 thanks!
@ptatian I think we can work with these files to re-create the master school file but the process will be different from the instructions.

ptatian commented:
@maggiesu0725 @rpitingolo One thing I would ask is that we create a permanent SAS data set on L: with the MSF (not just the summary files). That seems like it should be a natural output of the process. I don't care whether the data set is in the short or long format.

rpitingolo committed:
New master school file through 2017

rpitingolo committed:
fix ui_id numeric character issue

rpitingolo commented:
@ptatian I was able to get the MSF pretty close but need your assistance to proceed. It turns out that Enroll_15_16 has X and Y data but Enroll_16_17 and Enroll_17_18 do not. I was able to use proc ginside to join the 15_16 file to block and then use the %block10_to_geo macros to append the standard geographies. For the other two years I wasn't able to do this without the XY data. 

I looked around and if there is XY data for those years I don't know where it is. I could use the 15_16 file as a base and merge the XY data onto the other files. To do this we have to make the assumption that the UI_ID did not physically move. This also doesn't work if the UI_ID isn't in the 15_16 file. Another idea is to use the school name to manually geocode the remaining records after doing the merge. Does this seem like a reasonable approach? 

Including @maggiesu0725 in case she is online and knows where the XY data for Enroll_16_17 and Enroll_17_18 is located. Thanks.

maggiesu0725 commented:
@rpitingolo I don't have additional address information from my end, but I think we could get the latest school points with address and XY from DC open data and then use school name to link the two. I think using 15_16 file as a base also works, from the notes of experts (L:\Libraries\Schools\Raw\15-18 school update\notes_DC_school tracking_20180306) I don't see any existing schools moved (I could be wrong). I think merging XY in by UI_ID and use open data to fill in the gap of new schools is probably the easiest.  Just some thoughts!

rpitingolo commented:
@maggiesu0725 thanks. That's another idea. It looks like the files we have and the school file on Open Data both have a school ID that is different from UI_ID. I would need to confirm it matches but that would be preferable to merging by name, I think.

ptatian commented:
@rpitingolo @maggiesu0725 Sounds like a good plan to me. Let me know if you are still unsure how to proceed.

rpitingolo commented:
@ptatian thanks, Peter. I'm planning to wrap this up tomorrow.

rpitingolo committed:
Add manual geocodes

rpitingolo committed:
Add finalize macro

rpitingolo committed:
Add closed schools to newest years

rpitingolo committed:
flag closed schools in final file

rpitingolo commented:
@ptatian can you please review Master School Update Through 2017-18.sas ?

We currently don't have a school/year level Master School File registered in the warehouse/metadata. I called it "MSF_base" similar to the parcel_base file with the idea that it would update over time with the newest information. If you think it's worth created a dated version, we can do that too.

rpitingolo commented:
@maggiesu0725 FYI using the school files from DC Open Data didn't work. There are duplicate school IDs for school that have multiple campuses. We will need to sort that out if we plan to use that data for future master school file updates.

ptatian commented on Prog/Master School Update Through 2017-18.sas:
```diff
@@ -0,0 +1,226 @@
+
+
+/* Use DC Data Macros to convert blocks to standard geos */
+data Schools_00_14_geo;
```
This data step does not do anything because *GeoBlk2000* is missing for every record. This is because **Schools_00_14** contains only CLOSED school records.

rpitingolo replied:
@ptatian thanks. Schools_00_14 should not be keeping only closed schools. Looks like I introduced that error at the end when I was trying to add the closed schools on to the newer years. It's corrected now.

ptatian commented on Prog/Master School Update Through 2017-18.sas:
```diff
@@ -0,0 +1,226 @@
+	by ui_id; 
+run;
+
+proc sort data = enroll.Enroll_16_17 
```
**Enroll_16_17** and **Enroll_17_18** have *ui_id_n* values that are duplicated across different schools. This is not consistent with the structure of the previous MFS, where ID numbers are unique for each school. The duplicate ID values are

- *ui_id_n*=1021300 : "BRIDGES PCS" and "BRIGHTWOOD EC"
- *ui_id_n*=2100302 : "CESAR CHAVEZ PCS FOR PUBLIC POLICY - PARKSIDE HS" and "CESAR CHAVEZ PCS FOR PUBLIC POLICY - PARKSIDE MS"
- *ui_id_n*=2102403 : "DC PREP PCS - ANACOSTIA CAMPUS" and "DC PREP. PCS- BENNING MS"

These are all new schools that do not appear in the 2000-14 MFS. The different schools need to be assigned unique ID numbers.

rpitingolo replied:
The first one appears to be a typo that I simply corrected to the right number. The other two appear to be new campuses of existing charter schools. I used the guidance in the instructions to create new ui_ids for them.

ptatian commented on Prog/Master School Update Through 2017-18.sas:
```diff
@@ -0,0 +1,226 @@
+
+/* Combine old and new data together */
+data schools_00_18_combined;
+	set Schools_00_14_geo schools_15_18_geo;
```
This is not properly creating the entire new MFS because **Schools_00_14_geo** contains only CLOSED school records and, as a result, a lot of observations have been dropped. For example, see *ui_id*=1020700 (BENNING ELEMENTARY SCHOOL), which has dropped data prior to 2008, and *ui_id*=01X02700 (CHILD & FAMILY SERVICES), which is entirely missing from the new file. This needs to be fixed.

rpitingolo replied:
This should be corrected now that I resolved the issue from the first change request.

ptatian reviewed (changes requested):
@rpitingolo I finished my review of **Prog\Master School Update Through 2017-18.sas**. (I did not review the other programs as I'm assuming those are now obsolete? Let me know if that's not correct.) There are some issues that need to be fixed before the new file can be finalized.

rpitingolo committed:
schools_00_14 should keep all schools

Previous code had it keeping only closed schools. This was a typo and has been corrected.

rpitingolo committed:
Remove erroneous code in closed_schools data step

rpitingolo committed:
fix duplicate ui_id

rpitingolo commented:
@ptatian issues you identified should now be resolved if you want to take one final look before merging.

ptatian commented:
Thanks, @rpitingolo. I will look at it now.

ptatian committed:
Update Master School Update Through 2017-18.sas

Fix duplicate obs. Clean up formats. Add dup obs check.

ptatian commented:
@rpitingolo There seems to have been duplicate obs in the original **old.Msf_final_00_14_flip** data set for ui_id='1021300'. I added some code to fix this. I also added code at the end to check for duplicate obs and cleaned up some formats (eg, 0/1 vars should get the dyesno. format assigned to them).

There are several variables missing from the new MSF that were in the previous version: *school_name sch_address cluster2000 cluster_tr2000 anc2002 psa2004 ward2002 zip eor x_coord y_coord grade_min grade_max adult_flag open*. The older geographies we can perhaps omit, but I think that we want to have the grades and other info. 

I see you coded the corrected ID numbers into the program. That's fine, but isn't there a spreadsheet that is used for reconciling the ID numbers? We should make the corrections there, too.  cc: @maggiesu0725

rpitingolo commented:
@ptatian thanks. I added back all the old geographies since that's easy to do. 

I can add sch_address but it's going to be missing for for 2015 through 2017. The reason I dropped it was for that reason. Do you think it's worth adding back if it's only for the older years?

I can write a macro to get the min/max grades from the enrollment data and add that back on. 

And yes, I already  talked to @maggiesu0725 about updating that spreadsheet.

rpitingolo committed:
min/max grade macro

rpitingolo committed:
add old geos, min max grad macro call

rpitingolo committed:
add on x and y vars

rpitingolo committed:
Add school name, add labels

rpitingolo commented:
@ptatian one more question. x_coord and y_coord on the old data look like they are in meters. On the new data they are in decimal degrees. Do you think it's better to A) try to convert one to the other or B) create two sets of variables. I lean toward B since it's the path of lease resistance.

ptatian commented:
@rpitingolo I would include sch_address for now because I would imagine that we would eventually want to try to fill it in for 2015-2017? Even if we can't go back and fill in all the gaps, it seems like going forward we want to have the school addresses.

Our practice in other data sets has been to use x/y for the MD state plane coordinates (meters) and lat/lon for the projected coordinates (degrees). We should follow the same convention here. cc: @maggiesu0725

ptatian commented:
@rpitingolo @maggiesu0725 FYI, a small change is I dropped the format on ui_id. It seems unnecessary to have a format on this variable when the school name is already present in the data set.

rpitingolo committed:
separate lat/lon vars

rpitingolo committed:
lat long updates and labels

rpitingolo commented:
@ptatian thanks, yes, I saw that you removed the format ui_id. I think that's fine. I split out x_coord / y_coord and lat / lon. I think that wraps up the outstanding fixes. 

OK to finalize now?

ptatian commented:
@rpitingolo Yes, OK to finalize. Thanks.
